### PR TITLE
FIx `PytorchObjectDetector` loss gradient bug

### LINKED ISSUE #2237: Adversarial Patch Error in `PyTorchObjectDetector`
**Describe the bug**
When running the adversarial patch attack for the `PyTorchObjectDetector` and `PyTorchFasterRCNN`, an error can occur when using any PyTorch-specific preprocessors.

**To Reproduce**
Run the adversarial patch attack on the `PyTorchObjectDetector` while using a Faster R-CNN model. When using additional PyTorch-specific preprocessors, the error above will occur.

**Expected behavior**
The attacks should run without error. It can be easily fixed by adding the following if-condition 

```python
if x_tensor.is_leaf:
    x_tensor.requires_grad = True
else:
    x_tensor.retain_grad()
```

when setting the gradients to true. This if-condition is already applied for the `PyTorchYolo` detector, it just needs to be copied to the `PyTorchObjectDetector`.

**Screenshots**

**System information (please complete the following information):**
 - OS
 - Python version
 - ART version or commit number
 - TensorFlow / Keras / PyTorch / MXNet version

### PR DESCRIPTION
**Pull request by f4str**
# Description

Add the missing if-condition when setting gradients in the `PyTorchObjectDetector` estimator such that loss gradients can be properly calculated. This if-condition is copied from `PyTorchYolo`.

Fixes #2237 

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [x] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] Tests for `PyTorchObjectDetector` are unchanged
- [ ] Tests for `PyTorchFasterRCNN` are unchanged

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by beat-buesser — APPROVED** [review]
Hi @f4str Thank you very much for fixing this bug! The changes look good to me.
